### ISSUE TITLE
Alternative compression for MetaIndex

### ISSUE THREAD

**Issue by cmacdonald** [terrier-org/terrier-core#163]
Currently we use zlib deflate, which is slow. See associated discussion in https://github.com/terrier-org/pyterrier/issues/3#issuecomment-842164800

LZ4 and Zstd are likely to be faster.

**Comment by cmacdonald** [comment]
Results on Vaswani. Index on disk.

"Compressing" refers to Deflate.

### 1. When indexing docnos only:

```
-rw-r--r-- 1 root root 168K May 17 17:23 Dpassage_index-Compressing/data.meta.zdata
-rw-r--r-- 1 root root 168K May 17 17:24 Dpassage_index-LZ4/data.meta.zdata
-rw-r--r-- 1 root root 681K May 17 17:23 Dpassage_index-Uncompressed/data.meta.zdata
-rw-r--r-- 1 root root 224K May 17 17:23 Dpassage_index-Zstd/data.meta.zdata
```
![image](https://user-images.githubusercontent.com/620938/118530992-38470a80-b73d-11eb-8b26-6b7cf79cf8e7.png)


### 2. When indexing docnos and text.

```
-rw-r--r-- 1 root root 2.4M May 17 16:37 passage_index-Compressing/data.meta.zdata
-rw-r--r-- 1 root root 3.4M May 17 16:37 passage_index-LZ4/data.meta.zdata
-rw-r--r-- 1 root root 135M May 17 16:37 passage_index-Uncompressed/data.meta.zdata
-rw-r--r-- 1 root root 2.1M May 17 16:37 passage_index-Zstd/data.meta.zdata
```

#### (a) when retrieving docnos only

![image](https://user-images.githubusercontent.com/620938/118531103-5c0a5080-b73d-11eb-808c-c943b1a511ba.png)

#### (b) when retrieving docno + text

![image](https://user-images.githubusercontent.com/620938/118531202-7cd2a600-b73d-11eb-94c2-f6a04bc8c11e.png)


### Observations
 - compression helps with size but not speed when compressing docnos only. No overhead on mean response time. (Perhaps this could be improved by removing trailing 0 bytes)
 - LZ4 always better for speed, but results in markedly bigger files when text is present. 
 - deflate is comparable for docno only, but markedly slower when text is present.
 - Except for the (current default) deflate, the overheads for keeping text but not using it are 36-41% (for deflate, its 200%)

**Comment by cmacdonald** [comment]
Merged for 5.5 branch. Default compression will be new issue.

**Comment by cmacdonald** [comment]
Merged for 5.5
